### PR TITLE
core: Implement handling of text control input

### PR DESCRIPTION
**Pull request by n0samu**
This is a rebase and update of #2081. The same TODO items still apply, but those would be for a possible future PR. I've moved the left/right arrow key handling into the same code path as the rest of the text control events. ~~I'm marking this PR as a draft because I plan to implement pasting from the clipboard on web - that's the main feature still missing. In the meantime, please provide your reviews/feedback!~~ I've implemented clipboard pasting on web!

https://github.com/ruffle-rs/ruffle/assets/71368227/abf6f939-1f8d-4faf-ad75-829d4da7d8ad

Fixes the save/load menus in #4817, Fixes part of #1891.

### CODE REVIEW

File: core/src/events.rs
Comment (iwannabethedev) [inline core/src/events.rs:35]: Nit: Would a (maybe one-line) comment describing what the event means be helpful? I do not understand this part of the codebase, however, so that might be superfluous.

File: core/src/backend/ui.rs
Comment (iwannabethedev) [inline core/src/backend/ui.rs:162]: (Nit: Maybe insert a `TODO` or `FIXME` comment here?).

File: core/src/display_object/edit_text.rs
Comment (iwannabethedev) [inline core/src/display_object/edit_text.rs:1198]: I am not certain that I understand this part of the code. But, can this overflow? Like, if `max_chars.max(0)` can yield `1`, `text_len` (can the text be set by a script, thus bypassing the `max_chars`limit? https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/text/TextField.html#maxChars . Though I might be misunderstanding the code) can yield `10`, and `selection_len` can yield 0, that would give: `1 as usize - (10 - 0)` which would evaluate to something negative despite it being unsigned.

File: core/src/display_object/edit_text.rs
Comment (iwannabethedev) [inline core/src/display_object/edit_text.rs:1202]: (Nit: Would a (maybe one-line) comment to document this function be helpful?).

File: core/src/display_object/edit_text.rs
Comment (iwannabethedev) [inline core/src/display_object/edit_text.rs:1368]: (Nit: Would a (maybe one-line) comment to document this function be helpful?).

File: core/src/display_object/edit_text.rs
Comment (iwannabethedev) [reply core/src/display_object/edit_text.rs:1186]: Maybe:

```rust
/// The number of characters that currently can be added, considering `TextField.maxChars` constraint, current text length and current text selection.
```

?

File: core/src/events.rs
Comment (iwannabethedev) [inline core/src/events.rs:339]: (Nit: Is this `TODO` still relevant?).

File: web/src/lib.rs
Comment (iwannabethedev) [inline web/src/lib.rs:1767]: (Nit: Is this `TODO` still relevant?).

**Comment by iwannabethedev** [comment]
I have only done a fairly light or superficial code, since I only have a little bit of knowledge and understanding of Rust and this part of the codebase, but at a glance the changes look good to me. I think that someone else should review it as well, but apart from one non-nitpick question, I only have nitpicks. And I think after one more reviewer and after the non-nitpick question has been answered, that this would be ready for merging.

**Comment by iwannabethedev** [comment]
(For some reason two Linux builds failed to install dependencies in the continuous integration actions, which is unrelated to this PR - if that is just spurious failures, I imagine it could make sense to restart those jobs - or just wait until reruns of tests are triggered).

File: desktop/src/main.rs
Comment (iwannabethedev) [inline desktop/src/main.rs:916]: (Nit: Maybe add `// TODO: Extend this` considering the `TODO`s in https://github.com/ruffle-rs/ruffle/pull/2081 ?).

File: core/src/display_object/edit_text.rs
Comment (n0samu) [reply core/src/display_object/edit_text.rs:1198]: Oh good point, I didn't think of that

File: core/src/display_object/edit_text.rs
Comment (n0samu) [reply core/src/display_object/edit_text.rs:1216]: That's not the correct behavior

File: core/src/events.rs
Comment (n0samu) [reply core/src/events.rs:339]: That was from the original PR I rebased, so not 100% sure TBH

File: core/src/backend/ui.rs
Comment (n0samu) [reply core/src/backend/ui.rs:162]: No, this is the NullUiBackend, which doesn't have a clipboard

File: core/src/display_object/edit_text.rs
Comment (n0samu) [reply core/src/display_object/edit_text.rs:1186]: Perfect, I've added the comment (will push later)

File: core/src/display_object/edit_text.rs
Comment (n0samu) [reply core/src/display_object/edit_text.rs:1357]: I'm just doing exactly the same thing that's already done in the existing `text_input` function: https://github.com/ruffle-rs/ruffle/blob/1065662e845651bba173236b4a76b74d2881ce56/core/src/display_object/edit_text.rs#L1248-L1256

I don't know too much about this but I think if it's fine there it's fine here. And well, implementing whatever AVM2 feature/event may rely on this could be done in a separate PR, if that's indeed something that's missing (I have no idea)

File: core/src/events.rs
Comment (n0samu) [reply core/src/events.rs:339]: Looking again, yes it's still relevant because it's referring to the unimplemented control codes for Home, End, etc

**Comment by n0samu** [comment]
Thanks for catching the maxChars issue, that was pretty important! Fixed now. And I tried to improve the comments. I don't feel really confident in writing documentation for those functions you suggested, since I didn't write one of them, plus it takes a surprisingly long time to come up with good wording. So I think I'll leave those for this PR

**Comment by iwannabethedev** [comment]
I have only done a fairly light or superficial code, since I only have a little bit of knowledge and understanding of Rust and this part of the codebase, but at a glance the changes look very good to me, and I think this is ready to be merged after another reviewer has reviewed the changes.

File: .cargo/config.toml
Comment (relrelb) [inline .cargo/config.toml:9]: 1. For sanity, I suggest this flag to take effect just for Web builds, like it were on df11cd6a04369a839249df6dc3c86e88d29ccb99:
```toml
[target.wasm32-unknown-unknown]
rustflags=["--cfg=web_sys_unstable_apis"]
```
2. I wonder if this addition means that `$RUSTFLAGS` can be removed from here:
https://github.com/ruffle-rs/ruffle/blob/1065662e845651bba173236b4a76b74d2881ce56/web/packages/core/package.json#L22

File: core/src/backend/ui.rs
Comment (relrelb) [inline core/src/backend/ui.rs:19]: `clipboard_content()` can have an immutable `&self` recevier, right?

File: desktop/src/ui.rs
Comment (relrelb) [inline desktop/src/ui.rs:66]: nit: I think `.unwrap_or_default()` can be used here.

File: .cargo/config.toml
Comment (danielhjacobs) [reply .cargo/config.toml:9]: For number 2, it does not. These flags are overridden in `package.json`, and need to be until https://github.com/rust-lang/cargo/issues/10271 lands.

Comment (danielhjacobs) [reply .cargo/config.toml:9]: See https://github.com/ruffle-rs/ruffle/pull/5834

File: core/src/backend/ui.rs
Comment (n0samu) [reply core/src/backend/ui.rs:19]: Arboard's Clipboard.get_text function takes a mutable &self, so this doesn't seem to work

File: .cargo/config.toml
Comment (n0samu) [reply .cargo/config.toml:9]: I don't know how to apply that condition for just one flag while keeping the others for all targets. Can you tell me how?

Comment (danielhjacobs) [reply .cargo/config.toml:9]: Doesn't seem possible without duplicating the other flags: https://github.com/rust-lang/cargo/issues/5376

Comment (n0samu) [reply .cargo/config.toml:9]: Ok, I'll leave it as-is then

**Review by Toad06 — APPROVED** [review]

File: .cargo/config.toml
Comment (danielhjacobs) [reply .cargo/config.toml:9]: Since https://github.com/rust-lang/cargo/pull/12148 exists now, perhaps you can add a TODO that once it stabilizes, the clippy lints can be moved to Cargo.toml and this, as the one remaining rustflag, can be specified as `[target.wasm32-unknown-unknown]`

**Review by adrian17 — APPROVED** [review]

File: .cargo/config.toml
Comment (n0samu) [reply .cargo/config.toml]: I just checked and actually Clippy seems to try to compile our web code with all targets enabled, not just wasm. So we need the flag enabled for all targets unless we can tell Clippy not to do that. I've added the note about that new PR though.

**Comment by iwannabethedev** [comment]
I would guess that this is ready to be merged once rebased given adrian17's approval.

**Review by Dinnerbone — APPROVED** [review]

**Comment by allancoding** [comment]
Yay this was the exact game I was looking at that needed this, that is why I tried to do so in #9557 & #9521. I just don't know much about Rust I am still learning it. 

Thanks for doing it for me!

**Comment by n0samu** [comment]
Glad to help 😄

**Pull request merged**
